rootfs/scheduler/__init__.py

```diff
@@ -686,18 +697,24 @@ def resolve_state(self, pod):
         return states.get(pod_state, pod_state)
 
     def _get_port(self, image):
-        try:
-            image = self.registry + '/' + image
-            repo = image.split(":")
-            # image already includes the tag, so we split it out here
-            docker_cli = Client(version="auto")
-            docker_cli.pull(repo[0]+":"+repo[1], tag=repo[2], insecure_registry=True)
-            image_info = docker_cli.inspect_image(image)
-            port = int(list(image_info['Config']['ExposedPorts'].keys())[0].split("/")[0])
-        except Exception:
-            port = None
-
-        return port
+        # try thrice to find the port before raising exception as docker-py is flaky
+        for i in range(3):
```

Comment:
This could use the Python [retrying](https://pypi.python.org/pypi/retrying) library, which gives you a flexible `@retry` decorator. (I bet if it were added we'd find other places in the codebase that could use it as well.)